File: gestures.py
"""
Gesture detectors.

Each gesture is a function that takes a `Features` object and returns True
when the gesture is happening. Register it with @gesture(name, priority).

  * name      -> must match a key in config.GESTURE_MEMES
  * priority  -> higher wins when several gestures match at once

To add YOUR OWN gesture:
  1. Write a function `def my_thing(f): return <bool>` using f.blend(...),
     f.hands, f.face_point(...), and dist(a, b).
  2. Decorate it with @gesture("my_thing", priority=N).
  3. Add "my_thing": "my_file.gif" to config.GESTURE_MEMES.

Tips for tuning:
  * Run `python main.py --debug` to print live blendshape scores and hand
    positions so you can pick thresholds that fit your face/camera.
  * Blendshape names come from MediaPipe (jawOpen, mouthSmileLeft,
    eyeBlinkLeft, browOuterUpLeft, mouthPucker, eyeSquintLeft, ...).
  * All coordinates are 0..1. Divide pixel-ish distances by f.face_width to
    stay independent of how close you sit to the camera.
"""

from features import dist
import custom_gestures

# Registry populated by the @gesture decorator.
GESTURES = []  # list of (name, priority, fn)


def gesture(name, priority=0):
    def wrap(fn):
        GESTURES.append((name, priority, fn))
        return fn
    return wrap


def active_gesture(f):
    """Return the name of the highest-priority matching gesture, or None."""
    best = None
    best_priority = -1
    for name, priority, fn in GESTURES:
        try:
            if fn(f) and priority > best_priority:
                best, best_priority = name, priority
        except (IndexError, KeyError):
            # Missing landmark this frame — just skip the gesture.
            continue
    return best


# =========================================================================
#  EXAMPLE GESTURES  — delete/edit freely, these are just starting points.
# =========================================================================

@gesture("shock", priority=50)
def shock(f):
    """Mouth wide open (surprise/scream)."""
    return f.has_face and f.blend("jawOpen") > 0.45


@gesture("hands_up", priority=40)
def hands_up(f):
    """Both hands raised above the nose."""
    if f.num_hands < 2 or not f.has_face:
        return False
    nose_y = f.face_point(1)[1]
    return all(h.point(h.WRIST)[1] < nose_y for h in f.hands)


@gesture("shush", priority=35)
def shush(f):
    """One index fingertip pressed to the middle of the lips."""
    if not f.has_face or f.num_hands < 1:
        return False
    lips = f.face_point(13)  # upper-inner lip
    for h in f.hands:
        if dist(h.index_tip, lips) < 0.06:
            return True
    return False


@gesture("thinking", priority=30)
def thinking(f):
    """Index fingertip near the chin."""
    if not f.has_face or f.num_hands < 1:
        return False
    chin = f.face_point(152)
    for h in f.hands:
        if dist(h.index_tip, chin) < 0.09:
            return True
    return False


@gesture("wink", priority=20)
def wink(f):
    """One eye closed while the other stays open."""
    if not f.has_face:
        return False
    left = f.blend("eyeBlinkLeft")
    right = f.blend("eyeBlinkRight")
    return (left > 0.5 and right < 0.3) or (right > 0.5 and left < 0.3)


@gesture("smile", priority=10)
def smile(f):
    """Both mouth corners pulled up."""
    if not f.has_face:
        return False
    return f.blend("mouthSmileLeft") > 0.4 and f.blend("mouthSmileRight") > 0.4

File: web/static/js/gestures.js
// Port of gestures.py (built-in predicates) + custom_gestures.py (recorded
// template matching), plus the "highest priority wins" merge from
// active_gesture().

import { dist, HandIdx } from "./features.js";

// =========================================================================
//  BUILT-IN GESTURES — direct port of the @gesture predicates in gestures.py.
//  Keyed by name; priorities mirror the Python decorators / seed.py.
// =========================================================================

export const BUILTIN_FNS = {
  // Mouth wide open (surprise/scream).
  shock: (f) => f.has_face && f.blend("jawOpen") > 0.45,

  // Both hands raised above the nose.
  hands_up: (f) => {
    if (f.num_hands < 2 || !f.has_face) return false;
    const noseY = f.face_point(1)[1];
    return f.hands.every((h) => h.point(HandIdx.WRIST)[1] < noseY);
  },

  // One index fingertip pressed to the middle of the lips.
  shush: (f) => {
    if (!f.has_face || f.num_hands < 1) return false;
    const lips = f.face_point(13);
    return f.hands.some((h) => dist(h.index_tip, lips) < 0.06);
  },

  // Index fingertip near the chin.
  thinking: (f) => {
    if (!f.has_face || f.num_hands < 1) return false;
    const chin = f.face_point(152);
    return f.hands.some((h) => dist(h.index_tip, chin) < 0.09);
  },

  // One eye closed while the other stays open.
  wink: (f) => {
    if (!f.has_face) return false;
    const left = f.blend("eyeBlinkLeft");
    const right = f.blend("eyeBlinkRight");
    return (left > 0.5 && right < 0.3) || (right > 0.5 && left < 0.3);
  },

  // Both mouth corners pulled up.
  smile: (f) =>
    f.has_face &&
    f.blend("mouthSmileLeft") > 0.4 &&
    f.blend("mouthSmileRight") > 0.4,
};

// =========================================================================
//  RECORDED TEMPLATE MATCHING — port of custom_gestures.py.
// =========================================================================

const PTS_PER_HAND = 21;
export const DEFAULT_PRIORITY = 100; // recorded beats the built-in examples

// Build the numeric feature vector for frame `f` given an ordered name list.
// Layout: [blendshape scores] [Left: presence + 21 (x,y)] [Right: same],
// hand points measured relative to the nose and divided by face width.
export function featureVector(f, blendNames) {
  const vec = blendNames.map((n) => f.blend(n));

  const anchor = f.has_face ? f.face_point(1) : null; // nose tip
  const scale = f.face_width;

  const byLabel = { Left: null, Right: null };
  for (const h of f.hands) byLabel[h.label] = h;

  for (const label of ["Left", "Right"]) {
    const h = byLabel[label];
    if (h === null) {
      vec.push(0.0);
      for (let i = 0; i < 2 * PTS_PER_HAND; i++) vec.push(0.0);
      continue;
    }
    vec.push(1.0);
    const [ax, ay] = anchor ? anchor : h.point(HandIdx.WRIST);
    for (const [x, y] of h.landmarks) {
      vec.push((x - ax) / scale);
      vec.push((y - ay) / scale);
    }
  }
  return vec;
}

// Per-dimension weights matching featureVector's layout.
export function featureWeights(nBlend) {
  const w = new Array(nBlend).fill(1.0);
  for (let i = 0; i < 2; i++) {
    w.push(3.0); // hand presence matters a lot
    for (let j = 0; j < 2 * PTS_PER_HAND; j++) w.push(0.5);
  }
  return w;
}

export function wdist(a, b, w) {
  let s = 0.0;
  for (let i = 0; i < a.length; i++) {
    const d = a[i] - b[i];
    s += w[i] * d * d;
  }
  return Math.sqrt(s);
}

// =========================================================================
//  MATCHER — picks the highest-priority matching gesture for a frame.
//  Built with the subset of gestures "in play" for a session.
// =========================================================================

export class GestureMatcher {
  // `gestures` is an array of DB gesture docs: {id, name, kind, priority,
  // template}. Recorded docs carry template = {mean, threshold, blend_names}.
  constructor(gestures) {
    this.builtins = [];
    this.recorded = [];
    for (const g of gestures) {
      if (g.kind === "builtin") {
        const fn = BUILTIN_FNS[g.name];
        if (fn) this.builtins.push({ id: g.id, priority: g.priority, fn });
      } else if (g.template && g.template.mean) {
        this.recorded.push({
          id: g.id,
          priority: g.priority ?? g.template.priority ?? DEFAULT_PRIORITY,
          mean: g.template.mean,
          threshold: g.template.threshold,
          blendNames: g.template.blend_names,
          weights: featureWeights(g.template.blend_names.length),
        });
      }
    }
  }

  // Return the id of the winning gesture for frame `f`, or null.
  activeGesture(f) {
    let bestId = null;
    let bestPriority = -1;

    for (const b of this.builtins) {
      try {
        if (b.fn(f) && b.priority > bestPriority) {
          bestId = b.id;
          bestPriority = b.priority;
        }
      } catch {
        // Missing landmark this frame — just skip the gesture.
      }
    }

    // Closest recorded template within its threshold (custom_gestures.best_custom),
    // then folded into the same priority contest.
    let bestDist = Infinity;
    let recWinner = null;
    for (const r of this.recorded) {
      const vec = featureVector(f, r.blendNames);
      const d = wdist(vec, r.mean, r.weights);
      if (d <= r.threshold && d < bestDist) {
        bestDist = d;
        recWinner = r;
      }
    }
    if (recWinner && recWinner.priority > bestPriority) {
      bestId = recWinner.id;
      bestPriority = recWinner.priority;
    }

    return bestId;
  }

  // For tuning/debug: nearest recorded template as {id, dist, threshold}[].
  recordedScores(f) {
    return this.recorded
      .map((r) => ({
        id: r.id,
        dist: wdist(featureVector(f, r.blendNames), r.mean, r.weights),
        threshold: r.threshold,
      }))
      .sort((a, b) => a.dist - b.dist);
  }
}
